internal/core/model/adapters.rs

```diff
@@ -0,0 +1,282 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+//! This module contains adapter models.
+
+use super::*;
+
+pub struct MapModel<M, F> {
+    wrapped_model: M,
+    map_function: F,
+}
+
+impl<M, F, T, U> Model for MapModel<M, F>
+where
+    M: 'static,
+    F: 'static,
+    F: Fn(T) -> U,
+    M: Model<Data = T>,
+{
+    type Data = U;
+
+    fn row_count(&self) -> usize {
+        self.wrapped_model.row_count()
+    }
+
+    fn row_data(&self, row: usize) -> Option<Self::Data> {
+        self.wrapped_model.row_data(row).map(|x| (self.map_function)(x))
+    }
+
+    fn model_tracker(&self) -> &dyn ModelTracker {
+        self.wrapped_model.model_tracker()
+    }
+
+    fn as_any(&self) -> &dyn core::any::Any {
+        self
+    }
+}
+
+impl<M, F, T, U> MapModel<M, F>
+where
+    M: 'static,
+    F: 'static,
+    F: Fn(T) -> U,
+    M: Model<Data = T>,
+{
+    pub fn new(model: M, map_function: F) -> Self {
+        Self { wrapped_model: model, map_function }
+    }
+}
+
+#[test]
+fn test_map_model() {
+    let wrapped_rc = Rc::new(VecModel::from(vec![1, 2, 3]));
+    let map = MapModel::new(wrapped_rc.clone(), |x| x.to_string());
+
+    wrapped_rc.set_row_data(2, 42);
+    wrapped_rc.push(4);
+
+    assert_eq!(map.row_data(2).unwrap(), "42");
+    assert_eq!(map.row_data(3).unwrap(), "4");
+    assert_eq!(map.row_data(1).unwrap(), "2");
+}
+
+struct FilterModelInner<M, F>
+where
+    M: Model + 'static,
+    F: Fn(&M::Data) -> bool + 'static,
+{
+    wrapped_model: M,
+    filter_function: F,
+    // This vector saves the indices of the elements that are not filtered out
+    mapping: RefCell<Vec<usize>>,
+    notify: ModelNotify,
+}
+
+impl<M, F> FilterModelInner<M, F>
+where
+    M: Model + 'static,
+    F: Fn(&M::Data) -> bool + 'static,
+{
+    fn build_mapping_vec(&self) {
+        let mut mapping = self.mapping.borrow_mut();
+        *mapping = self
+            .wrapped_model
+            .iter()
+            .enumerate()
+            .filter(|(_, e)| (self.filter_function)(e))
+            .map(|(i, _)| i)
```

Comment:
```suggestion
            .filter_map(|(i, e)| (self.filter_function)(&e).then(|| i))
```

(untested)

Reply:
> > Then we have to clone() the element all the time when we check it with the filter_function because it is consuming.
> 
> I don't think you would need to clone. You would have to use the `Iterator::map_filter` function and forward the element to the filter function.

Oh right. But for the row_changed implementation we still would have to copy.
That is not a big problem but actually now that I know that iterator map and filter behave the same, I actually want to keep it like it is. This way it will be more familiar to users.

Reply:
That is pretty cool. I didn't know about bool's then().